### PR TITLE
Add NeXus friendly fields to DCM

### LINKED ISSUE #397: Port DCM to ophyd-async
Required for #377

I22 has a DCM, so can port the existing implementation and test it works for all beamlines. The linked issue shows the extra fields the device would need to expose.

## Acceptance Criteria
* ophyd-async DCM device available
* DCM configuration included in i22 beamline module
* Fake DCM configuration included p38 beamline module
* Fields specified in #377 are exposed

### PR DESCRIPTION
**Pull request by callumforrester**
Fixes #397 

### Instructions to reviewer on how to test:
1. Confirm tests pass

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://github.com/DiamondLightSource/dodal/wiki/Device-Standards)
- [ ] If changing the API for a pre-existing device, ensure that any beamlines using this device have updated their Bluesky plans accordingly

### CODE REVIEW

File: src/dodal/beamlines/i22.py [hunk 1]
```diff
@@ -120,6 +121,25 @@ def hfm(
     )
 
 
+def dcm(
+    wait_for_connection: bool = True,
+    fake_with_ophyd_sim: bool = False,
+) -> FocusingMirror:
```
Comment (DiamondJoseph) [inline src/dodal/beamlines/i22.py:127]:
```suggestion
) -> DCM:
```

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]
Bad News: NXinsertion_device can take arbitrarily many crystals, and a DualCrystalMonochromator has two crystals.

The i22 requirement has

crystal_1 = DCMCrystal(
    usage="Bragg",
    type="Silicon",
    reflection=(1, 1, 1),
    d_spacing=3.13475,
)

crystal_2 = DCMCrystal(
    usage="Bragg",
    type="Silicon",
    reflection=(1, 1, 1),
    d_spacing=3.13475,
)

I think it'll need to take a dict...

File: src/dodal/devices/dcm.py [hunk 2]
```diff
@@ -17,6 +33,7 @@ def __init__(
         self,
         prefix: str,
         name: str = "",
+        crystal: DCMCrystal | None = None,
```
[the remark this answers is not in the document]
Comment (callumforrester) [reply src/dodal/devices/dcm.py:36]: Hmm, do we need to wait for this then?

https://github.com/bluesky/ophyd-async/issues/311

File: src/dodal/devices/dcm.py [hunk 3]
```diff
@@ -1,8 +1,24 @@
...
+from dataclasses import dataclass
+from typing import Literal
+
+from ophyd_async.core import ConfigSignal, StandardReadable, soft_signal_r_and_setter
 from ophyd_async.epics.motion import Motor
 from ophyd_async.epics.signal import epics_signal_r
 
 
+@dataclass
+class DCMCrystal:
```
Comment (callumforrester) [inline src/dodal/devices/dcm.py:10]:
```suggestion
class MonochromatorCrystal:
```

**Comment by callumforrester** [comment]
Currently fails the system test for i22, PV prefix may be wrong

**Comment by callumforrester** [comment]
Update on this issue: I think for now we need to make another DCM class and an issue to reconcile. 
Supporting multiple crystals is also still TODO

**Comment by callumforrester** [comment]
Closing for now in favour of #598

**Pull request closed without merging**